### DISCUSSION TITLE
Does Nuxt 3 have transitions?

### DISCUSSION THREAD

**Discussion by joewearing** [nuxt/nuxt#16183] [Questions]
In Nuxt 2 we had [page transitions](https://nuxtjs.org/docs/features/transitions/)

Does this / will this exist in Nuxt 3, the same transition property doesn't seem to do anything

Being able to add my own `<transition>` component within `vue.app` (around `<NuxtPage>`) would also be helpful!

**Comment by lucpotage** [comment]
If I’m not wrong, transitions are made possible thanks to `vue-router`. [In the docs](https://v3.nuxtjs.org/docs/directory-structure/app), it’s mentioned that when using a single `app.vue` file, `vue-router` dependency is NOT included to reduce the bundle size for simple apps. Did you try to create a `pages/index.vue` file instead?

**Comment by maxdzin** [comment]
I've used Nuxt 3 layout + pages.
In `nuxt.config.ts` I've specified `pageTransition: {'page-fade': true },` and then specified a transition styles as of https://v3.vuejs.org/guide/transitions-enterleave.html

If you will take a look at Nuxt config parameters, you can see that there are `pageTransitions`, `layoutTransitions` are available to handle that in several ways for your application.
![image](https://user-images.githubusercontent.com/6406412/139430397-a1fd6eb4-39cb-4c4e-9d89-dee759b035e0.png)

**Reply by jonmes** [reply]
how can we do it with javascript hooks, like in case of using gsap.
I tried it but the hooks are not triggered without the condition or (v-if), it should work without v-if when element is mounted for the first time, I tried the same exact code on vue 3 and it works but not with nuxt 3.

code is here for Nuxt 3
```
<script setup>
import gsap from "gsap";

const images = [
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
];

const beforeEnter = (el) => {
  console.log("beforeEnter ,");
  el.style.transform = "translateY(100px)";
  el.style.opacity = 0;
};

const enter = (el, done) => {
  console.log("enter");
  gsap.to(el, {
    y: 0,
    opacity: 1,
    duration: 0.7,
    onComplete: done,
    delay: el.dataset.index * 0.1,
  });
};
</script>

<template>
  <div class="tran">
    <transition-group
      appear="true"
      @before-enter="beforeEnter"
      @enter="enter"
      name="stagger"
      tag="ul"
    >
      <li v-for="(img, i) in images" class="list" :key="i" :data-index="i">
        <img alt="Vue logo" :src="img" style="width: 100px" />
      </li>
    </transition-group>
  </div>
</template>
```

**Comment by Zizaco** [comment]
@joewearing you can use standard vue transitions **inside** your components:

```vue
<template>
  <transition name="foobar">
    <div v-if="error">
      <strong>{{ error }}</strong>
    </div>
  </transition>
</template>

<!-- script -->

<style scoped>
.foobar-enter-from{
  opacity: 0;
  transform: translateY(15px);
}

.foobar-enter-active,
.foobar-leave-active {
  transition: all 0.5s;
}

.foobar-enter,
.foobar-leave-to {
  opacity: 0;
}
</style>
```

Concerning **page transitions**, currently your are limited to a single transition to the whole app, called `"page"`. Basically because you can't yet [change its name like you could in nuxt2](https://nuxtjs.org/docs/features/transitions/#string).

So if you include the following in your css, it should add a basic transition to your pages:

```css
/* "page" is hardcoded in nuxt3 page transitions atm */
.page-enter-from{
  opacity: 0;
  transform: translateY(-10px);
}

.page-enter-active,
.page-leave-active {
  transition: all 0.2s;
}
.page-enter,
.page-leave-to {
  opacity: 0;
}
```

The `name="page"` is hardcoded on `nuxt3/src/pages/runtime/page.vue` atm. [See source](https://github.com/nuxt/framework/blame/main/packages/nuxt3/src/pages/runtime/page.vue#L4).

I believe that nuxt team will be adding transition names to pages soon.

**Reply by danielroe** [reply]
For future reference, this does now exist: see https://v3.nuxtjs.org/docs/directory-structure/pages#layouttransition-and-pagetransition

**Reply by floriankapaun** [reply]
Just to have it mentioned, it is currently not possible to use conditional names for a pageTransition at the moment.

**Reply by jonmes** [reply]
how can we do it with javascript hooks, like in case of using gsap.
I tried it but the hooks are not triggered without the condition or (v-if), it should work without v-if when element is mounted for the first time, I tried the same exact code on vue 3 and it works but not with nuxt 3.

code is here for Nuxt 3
```
<script setup>
import gsap from "gsap";

const images = [
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
  "https://cdn.pixabay.com/photo/2022/06/02/19/25/giraffes-7238815_960_720.jpg",
];

const beforeEnter = (el) => {
  console.log("beforeEnter ,");
  el.style.transform = "translateY(100px)";
  el.style.opacity = 0;
};

const enter = (el, done) => {
  console.log("enter");
  gsap.to(el, {
    y: 0,
    opacity: 1,
    duration: 0.7,
    onComplete: done,
    delay: el.dataset.index * 0.1,
  });
};
</script>

<template>
  <div class="tran">
    <transition-group
      appear="true"
      @before-enter="beforeEnter"
      @enter="enter"
      name="stagger"
      tag="ul"
    >
      <li v-for="(img, i) in images" class="list" :key="i" :data-index="i">
        <img alt="Vue logo" :src="img" style="width: 100px" />
      </li>
    </transition-group>
  </div>
</template>
```

**Comment by ka3hun9** [comment]
` <Transition name="layout">` and `<Transition name="page">` exists natively in Nuxt3. You just need to provide the global css ;

like this:
`.layout-enter-active,
.layout-leave-active 
.page-enter-active,
.page-leave-active`

**Reply by fabianbeier** [reply]
Is there any way at the moment to access the JavaScript hooks to the global page transition? 

I tried setting them in the nuxt.config.ts through the pageTransition object but they did not fire.

**Reply by maxdzin** [reply]
How about this one?
```ts
const nuxtApp = useNuxtApp()
  
nuxtApp.hook('page:start', () => {
  console.log('page load start')
})
nuxtApp.hook('page:finish', () => {
  console.log('page load finish')
})
```

**Reply by atinux** [reply]
Nice answer @maxdzin 

We do need to add examples and documentation for page transitions, any help is welcome 💚

**Reply by maxdzin** [reply]
> Nice answer @maxdzin 
> 
> 
> 
> We do need to add examples and documentation for page transitions, any help is welcome 💚 

Ok. I'll try to check what can I do for that.
And thank you for such a great tool!

**Reply by dennishelbig** [reply]
If you are using `<NuxtPage /> ` to display your pages, you can still use `<router-view></router-view>` instead.

replace 
```html 
<NuxtPage /> 
```
 with
```html
<router-view v-slot="{ Component }">
  <transition 
    mode="out-in"
    @enter="enterFunction"
    @leave="leaveFunction">
      <component :is="Component" />
  </transition>
</router-view>
```
( No need to import anything, it should work by itself )

In your methods, you can now execute JS like in vue 3.
for example:

```javascript
export default {
  methods: {
    leaveFunction( page, done ){
      console.log(page) // do something with page
      done() // let nuxt know animation is finished
    },
    enterFunction( page, done ){
      console.log(page) // do something with page
      done() // let nuxt know animation is finished
    }
  }
}
```

I'm using Nuxt3 rc4.

**Reply by letoast** [reply]
> If you are using `<NuxtPage /> ` to display your pages, you can still use `<router-view></router-view>` instead.
> 
> 
> I'm using Nuxt3 rc4.

Using your method "by just replacing" <NuxtPage />, the page is empty after navigation. It is passed to those 2 functions, but it's empty. I'm using script setup, composition API

**Comment by lacorde** [comment]
Is it possible do disable global page transition for specific routes or pages ?

**Reply by gluktd** [reply]
```
definePageMeta({
  pageTransition: false
})
```

**Comment by sillvva** [comment]
It is possible to apply a transition only to the slots on a layout, while the other elements of the layout are fixed?

**Reply by lucpotage** [reply]
Have you tried https://github.com/antfu/vue-starport? It may help.

**Reply by sillvva** [reply]
I have not, but I honestly don't care if the fixed elements are destroyed and recreated. I just don't want them to be animated as part of the transition. I only want the transition to apply to certain elements of the layout.

**Comment by oleg-savenok** [comment]
Is it possible to use transition events on `<Transition name="layout">` and `<Transition name="page">` like it shows in the Vue documentation?

https://vuejs.org/guide/built-ins/transition.html#javascript-hooks

```
<Transition
  @before-enter="onBeforeEnter"
  @enter="onEnter"
  @after-enter="onAfterEnter"
  @enter-cancelled="onEnterCancelled"
  @before-leave="onBeforeLeave"
  @leave="onLeave"
  @after-leave="onAfterLeave"
  @leave-cancelled="onLeaveCancelled"
>
  <!-- ... -->
</Transition>
```

**Reply by BenjaminOddou** [reply]
I tried with this config but it doesn't seems to work with nuxt 3. I also tried to use nuxt 3 hooks `page:start` and `page:finish` inside a plugin but I didn't found a way to synchronize them with a transition. For now, we are probably stuck with CSS transition I believe. 😕

**Reply by fabianbeier** [reply]
In your app.vue you can use the standard `<router-view>` element as documented [here](https://router.vuejs.org/guide/advanced/transitions.html)

```
<template>
  <div>
    <router-view v-slot="{ Component }">
      <transition @leave="onLeave" @enter="onEnter" mode="out-in" :css="false">
        <component :is="Component" />
      </transition>
    </router-view>
  </div>
</template>
```

Here is also a Stackblitz with a [demo](https://stackblitz.com/edit/nuxt-starter-j5pufo?file=app.vue). Hope this helps.

**Reply by oleg-savenok** [reply]
@fabianbeier, Well looks good, but what about Nuxt SSR? Is it going to work properly if just replace `<NuxtLayout> `and `<NuxtPage>` with standard Vue tags?

**Reply by BenjaminOddou** [reply]
Hello @fabianbeier and thank you very much for the demo ! I tried to play around what your provided but I couldn't find a way to make my transition works in [my demo](https://stackblitz.com/edit/nuxt-starter-ugbwzw?file=app.vue). I have this error popping 😕

![image](https://user-images.githubusercontent.com/85166574/181231466-258c389d-ed56-4c34-a9fa-bcb6844e2ce2.png)

**Reply by fabianbeier** [reply]
This has nothing to do with Nuxt. The functions „onEnter“ and „onLeave" need „done“ as the second argument as described in the Vue docs.

![image](https://user-images.githubusercontent.com/38300500/181235589-c1a6ae60-be80-4665-8b8b-71eb569be7e2.png)

**Reply by BenjaminOddou** [reply]
@fabianbeier do you think it's possible to build two or more different js transitions (based on your model) that are triggered depending on the button clicked ?  🧐

**Comment by EmilyFlocc** [comment]
Is it possible to have a different page transition on different pages? Ie. if I'm going from a parent page to a child page, I want to the child page to swipe in from the left. But if you navigate to any other page, I want it to fade between pages. 

Anyone know how I can achieve this logic in Nuxt 3? I've tried naming a transition in definePageMeta but it errors

`definePageMeta({
  pageTransition: 'swipe-up'
})`

`[nitro] [dev] [unhandledRejection] TypeError: Cannot use 'in' operator to search for '__vInternal' in swipe-up`

**Reply by akshay-sood** [reply]
Did you ever find solution to this? I am also hitting the same error :(

**Reply by EmilyFlocc** [reply]
No I haven't, still hoping someone will get back to me!

**Reply by maxdzin** [reply]
@EmilyFlocc take a look on definition of the `PageMeta`:
```ts
export interface PageMeta {
    [key: string]: any;
    pageTransition?: boolean | TransitionProps;
    layoutTransition?: boolean | TransitionProps;
    key?: false | string | ((route: RouteLocationNormalizedLoaded) => string);
    keepalive?: boolean | KeepAliveProps;
}
```

**pageTransition** can be `boolean` or `TransitionProps`:
```ts
export declare interface TransitionProps extends BaseTransitionProps<Element> {
    name?: string;
    type?: typeof TRANSITION | typeof ANIMATION;
    ...
}
```

So, you should specify the transition name for the page like this:
```ts
definePageMeta({ pageTransition: { name: 'swipe-up' } })
```

I'm using this way.
Nuxt v3 rc.11.
